pkg/target/http.go

```diff
@@ -215,63 +337,150 @@ func AdaptHTTPTargetFunc(f func(c *HTTPTargetConfig) (*HTTPTarget, error)) HTTPT
-		resp, err := ht.client.Do(request) // Make request
-		requestFinished := time.Now().UTC()
+	for _, chunk := range chunks {
+		grouped := ht.groupByDynamicHeaders(chunk)
 
-		msg.TimeRequestStarted = requestStarted
-		msg.TimeRequestFinished = requestFinished
+		for _, group := range grouped {
+			var reqBody []byte
+			var goodMsgs []*models.Message
+			var badMsgs []*models.Message
+			var err error
 
-		if err != nil {
-			errResult = multierror.Append(errResult, err)
-			failed = append(failed, msg)
-			continue
-		}
-		defer resp.Body.Close()
-		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
-			sent = append(sent, msg)
-			if msg.AckFunc != nil { // Ack successful messages
-				msg.AckFunc()
+			if ht.requestTemplate != nil {
+				reqBody, goodMsgs, badMsgs = ht.renderBatchUsingTemplate(group)
+			} else {
+				reqBody, goodMsgs, badMsgs = ht.provideRequestBody(group)
 			}
-		} else {
-			errResult = multierror.Append(errResult, errors.New("Got response status: "+resp.Status))
-			failed = append(failed, msg)
-			continue
+
+			invalid = append(invalid, badMsgs...)
+
+			if len(goodMsgs) == 0 {
+				continue
+			}
+
+			request, err := http.NewRequest("POST", ht.httpURL, bytes.NewBuffer(reqBody))
+
+			if err != nil {
+				panic(err)
+			}
+
+			request.Header.Add("Content-Type", ht.contentType)                        // Add content type
+			addHeadersToRequest(request, ht.headers, ht.retrieveHeaders(goodMsgs[0])) // Add headers if there are any - because they're grouped by header, we just need to pick the header from one message
+			if ht.basicAuthUsername != "" && ht.basicAuthPassword != "" {             // Add basic auth if set
+				request.SetBasicAuth(ht.basicAuthUsername, ht.basicAuthPassword)
+			}
+			requestStarted := time.Now()
+			resp, err := ht.client.Do(request) // Make request
+			requestFinished := time.Now()
+
+			// Add request times to every message
+			for _, msg := range goodMsgs {
+				msg.TimeRequestStarted = requestStarted
+				msg.TimeRequestFinished = requestFinished
+			}
+
+			if err != nil {
+				failed = append(failed, goodMsgs...)
+				errResult = multierror.Append(errResult, errors.New("Error sending request: "+err.Error()))
+				continue
+			}
+			defer resp.Body.Close()
+
+			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+				for _, msg := range goodMsgs {
+					if msg.AckFunc != nil { // Ack successful messages
+						msg.AckFunc()
+					}
+					sent = append(sent, msg)
+				}
+				continue
+			}
+
+			responseBody, err := io.ReadAll(resp.Body)
+
+			if err != nil {
+				failed = append(failed, goodMsgs...)
+				errResult = multierror.Append(errResult, errors.New("Error reading response body: "+err.Error()))
+				continue
+			}
+
+			response := response{Body: string(responseBody), Status: resp.StatusCode}
+
+			if findMatchingRule(response, ht.responseRules.Invalid) != nil {
+				for _, msg := range goodMsgs {
+					// can we use response body as an error message for invalid data?
```

Comment:
nit: I don't think this comment is intended for production.

AFAIR we settled the question itself - I believe we landed on "as long as it's not getting dumped to logs, then yes it can be provided to the failure target". :)